### PR TITLE
Feat/create bid

### PR DESCRIPTION
**Pull request by andrew2003**
- [ ] Create an issue explaining what tasks should be done by this PR, diagrams, etc.:  https://github.com/sota-zk-labs/tahm-kench/issues

### CODE REVIEW

File: contracts/src/ZkAuction.sol [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        bytes encryptedPrice; // Encrypted price submitted of bidder
+    }
+    struct Asset {
+        string name; // Name of the asset
+        string description; // Description of the asset
+    }
+
+    struct Winner {
+        address winner; // Address of the winner
+        bytes encryptedPrice; // Encrypted price submitted of winner
```
Comment (0x5ea000000) [inline contracts/src/ZkAuction.sol]: why the price is still encrypted here? it should be unencrypted price

File: contracts/src/ZkAuction.sol [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    function deposit(uint256 auction_id) public {
+        Auction storage auction = auctions[auction_id];
+        require(!auction.hasDeposited[msg.sender], "Already deposited");
+        require(!auction.ended, "Auction ended");
+        // Transfer tokens from the user to this contract
+        lockToken.transferFrom(msg.sender, address(this), auction.depositPrice);
+        // Update the state to indicate that the user has deposited
+        auction.hasDeposited[msg.sender] = true;
```
Comment (0x5ea000000) [inline contracts/src/ZkAuction.sol]: update all contract inner state before calling to external contract

File: src/main.rs [hunk 3]
```diff
@@ -1,3 +1,161 @@
...
+        private_key: Bytes,
+    },
+    /// Claim item
+    Claim {
+        #[arg(short, long)]
+        id: i32,
+    },
+}
+
+#[tokio::main]
```
Comment (0x5ea000000) [inline src/main.rs]: check other rust repo to understand how to structure code base

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,7 @@
+SHELL := /bin/bash
+
+cargo-run:
+	@cargo run --release
+	@cargo install --path .
```
Comment (zk-steve) [inline Makefile]: ?

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+deploy-contract:
+	@. ./contracts/.env && . ./contracts/deploy.sh
```
Comment (zk-steve) [inline Makefile]: don't inject .env here

```suggestion
	@. ./contracts/deploy.sh
```

File: config.toml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+contract_address = "0xaa80ab74e426a0d2c19178db78649eebd05d05c5"
+[chain]
```
Comment (zk-steve) [inline config.toml:2]:
```suggestion

[chain]
```

File: contracts/.github/workflows/test.yml [hunk 7]
```diff
@@ -0,0 +1,45 @@
+name: CI
+
+on:
+  push:
```
Comment (zk-steve) [inline contracts/.github/workflows/test.yml]:
```suggestion
```

File: contracts/script/Deployer.s.sol
Comment (zk-steve) [inline contracts/script/Deployer.s.sol:1]: file name convention

File: contracts/src/ZkAuction.sol
Comment (zk-steve) [inline contracts/src/ZkAuction.sol:1]: file name convention

File: src/assets/ZkAuction.json
Comment (zk-steve) [inline src/assets/ZkAuction.json:1]: file name convention

File: contracts/test/ZkAuction.t.sol
Comment (zk-steve) [inline contracts/test/ZkAuction.t.sol:1]: file name convention

File: src/config.rs [hunk 8]
```diff
@@ -0,0 +1,27 @@
+use config::{Config as ConfigLoader, File, FileFormat};
+use ethers::types::H160;
+use serde::Deserialize;
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct ChainConfig {
+    pub rpc_url: String,
+    // pub etherscan_api_key: String
+    // pub network:
```
Comment (zk-steve) [inline src/config.rs]:
```suggestion
```

File: src/main.rs [hunk 9]
```diff
@@ -1,3 +1,161 @@
...
+use ethers::providers::{Http, Provider};
+use ethers::signers::{LocalWallet, Signer};
+use ethers::types::Bytes;
+
+use crate::auction::{create_new_auction, get_auction, get_total_auction};
+
+mod auction;
+mod config;
+
+/// TAHKEN
```
Comment (zk-steve) [inline src/main.rs]: name ?.

File: src/main.rs [hunk 10]
```diff
@@ -1,3 +1,161 @@
...
+    command: Option<Commands>,
+    #[clap(short, long)]
+    version: bool,
+    #[clap(short, long, default_value = "config.toml")]
+    config_path: String,
+    /// Path of local wallet
+    #[clap(
+        short,
+        long,
+        default_value = "/home/ubuntu/.foundry/keystores/wallet_zk_auction"
```
Comment (zk-steve) [inline src/main.rs]: not all ubuntu os have this path

File: src/main.rs [hunk 11]
```diff
@@ -1,3 +1,161 @@
...
+    let private_key = wallet.signer();
+    let public_key = private_key.verifying_key();
+    // Convert the public key to Bytes
+    let public_key_bytes = Bytes::from(public_key.to_encoded_point(false).as_ref().to_vec());
+
+    let signer = SignerMiddleware::new(provider.clone(), wallet.clone());
+
+    match args.command {
+        Some(command) => match command {
+            Commands::Version => {
```
Comment (zk-steve) [inline src/main.rs]: correct way, https://github.com/zk-steve/rust-web-api-microservice-template/blob/main/src/public/src/main.rs#L32

File: src/main.rs [hunk 12]
```diff
@@ -1,3 +1,161 @@
...
+        description: String,
+        #[arg(short, long)]
+        target_price: U256,
+        #[arg(short, long, default_value = "1")]
+        time: u64,
+    },
+    /// Get list auctions opening
+    ListAuctions,
+    /// Get detail auctions
+    GetAuctions {
```
Comment (zk-steve) [inline src/main.rs]:
```suggestion
    GetAuction {
```

File: contracts/src/ZkAuction.sol (hunk 1 above)
Comment (andrew2003) [reply contracts/src/ZkAuction.sol]: Done !

File: config.toml (hunk 6 above)
Comment (andrew2003) [reply config.toml:2]: Done

File: contracts/.github/workflows/test.yml (hunk 7 above)
Comment (andrew2003) [reply contracts/.github/workflows/test.yml]: Done

File: src/config.rs (hunk 8 above)
Comment (andrew2003) [reply src/config.rs]: Done

File: src/main.rs (hunk 12 above)
Comment (andrew2003) [reply src/main.rs]: Done

File: src/main.rs (hunk 10 above)
Comment (andrew2003) [reply src/main.rs]: Done

File: contracts/script/Deployer.s.sol
Comment (andrew2003) [reply contracts/script/Deployer.s.sol:1]: Done

File: contracts/src/ZkAuction.sol
Comment (andrew2003) [reply contracts/src/ZkAuction.sol:1]: Done

File: contracts/test/ZkAuction.t.sol
Comment (andrew2003) [reply contracts/test/ZkAuction.t.sol:1]: Done

File: src/assets/ZkAuction.json
Comment (andrew2003) [reply src/assets/ZkAuction.json:1]: Done

File: Makefile [hunk 13]
```diff
@@ -0,0 +1,16 @@
+SHELL := /bin/bash
+
+CONFIG_PATH=config.toml
+
+start-cli:
+	@cargo install --bin tahken --path .
+	@tahken \
+		--config-path $(CONFIG_PATH) \
+		--keystore-path $(KEYSTORE_PATH)
+	@rm -rf ~/.zk_auction/keystores/wallet_zk_auction
```
Comment (0x5ea000000) [inline Makefile]: the config folder name should be the same with the cli name

File: LICENSE [hunk 14]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) [2024] [Andrew Tran]
```
Comment (zk-steve) [inline LICENSE]:
```suggestion
Copyright (c) [2024]
```

File: Makefile [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+start-cli:
+	@cargo install --bin tahken --path .
+	@tahken \
+		--config-path $(CONFIG_PATH) \
+		--keystore-path $(KEYSTORE_PATH)
+	@rm -rf ~/.tahken/keystores/wallet_tahken
+	@mkdir -p ~/.tahken/keystores
+	@cp $(KEYSTORE_PATH) ~/.tahken/keystores
+
+
```
Comment (zk-steve) [inline Makefile:7]:
```suggestion
```

File: src/main.rs (hunk 9 above)
Comment (andrew2003) [reply src/main.rs]: Tahm-kench is the project name, Tahken is the abbreviation

File: src/bin/tahken.rs [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+
+    let keystore_password = rpassword::prompt_password("Enter keystore password: ")
+        .expect("Failed to read keystore password");
+
+    // let home_dir = home_dir().expect("Failed to get home directory");
+    // let path = home_dir.join(&args.keystore_path);
+    //
+    // let wallet = LocalWallet::decrypt_keystore(path, &keystore_password)
+    //     .expect("Failed to decrypt keystore")
+    //     .with_chain_id(chain_id.as_u64());
```
Comment (zk-steve) [inline src/bin/tahken.rs]: remove

File: sp1-prover/src/main.rs [hunk 17]
```diff
@@ -0,0 +1,27 @@
+#![no_main]
```
Comment (zk-steve) [inline sp1-prover/src/main.rs]: ?

File: sp1-prover/src/main.rs [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+        let bidder_amount = decrypt_bidder_data(&pvk, bidder);
+        if winner_amount < bidder_amount {
+            winner_amount = bidder_amount;
+            winner_addr = &bidder.address;
+        }
+    }
+
+    sp1_zkvm::io::commit(&calc_auction_hash(&auction_data));
+    sp1_zkvm::io::commit(winner_addr);
+    sp1_zkvm::io::commit(&winner_amount);
```
Comment (zk-steve) [inline sp1-prover/src/main.rs]: why `&` here ?

File: sp1-prover/src/main.rs [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+pub fn main() {
+    let auction_data = sp1_zkvm::io::read::<AuctionData>();
+
+    let pvk = SecretKey::parse_slice(&sp1_zkvm::io::read::<Vec<u8>>())
+        .expect("missing private key to encode bidder data");
+
+    let mut winner_addr = &vec![];
+    let mut winner_amount = 0;
+    for bidder in &auction_data.bidders {
+        let bidder_amount = decrypt_bidder_data(&pvk, bidder);
```
Comment (zk-steve) [inline sp1-prover/src/main.rs]: Can decrypting only bidder amount improve the prove time?
`let bidder_amount = decrypt_bidder_data(&pvk, bidder.amount)`

File: sp1-prover/src/main.rs [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+        if winner_amount < bidder_amount {
+            winner_amount = bidder_amount;
+            winner_addr = &bidder.address;
+        }
+    }
+
+    sp1_zkvm::io::commit(&calc_auction_hash(&auction_data));
```
Comment (zk-steve) [inline sp1-prover/src/main.rs]: why reference here?

File: sp1-prover/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+        fs::write("encryption_key", pbk).expect("failed to write public key to file");
+    }
+
+    #[test]
+    fn test_hash_auction() {
+        let (_, pbk) = get_key();
+        let data = auction_data(&pbk);
+        println!("{:?}", data);
+        // println!("{:?}", hex::encode(&data.bidders[0].address));
+        // println!("{:?}", hex::encode(&data.bidders[1].encrypted_amount));
```
Comment (zk-steve) [inline sp1-prover/src/lib.rs]: remove

File: sp1-prover/Makefile [hunk 22]
```diff
@@ -0,0 +1,11 @@
+test:
+	RUST_BACKTRACE=1 cargo test --release --color=always --package aligned-sp1-prover --lib tests::test_sp1_prover --no-fail-fast -- --exact -Z unstable-options --show-output
+	
+gen-key:
+	touch encryption_key
+	touch private_encryption_key
```
Comment (zk-steve) [inline sp1-prover/Makefile]: make the program handle this instead

File: config.toml (hunk 6 above)
Comment (zk-steve) [reply config.toml:2]: >?

Comment (zk-steve) [reply config.toml:2]: ?

File: config.toml [hunk 23]
```diff
@@ -0,0 +1,10 @@
+contract_address = "0x07e45d6971dd3e72f9a9408560554754462ab807"
+token_address = "0x9021e645c961993d2b91042345d7a0e125452ed0"
+nft_address = "0x4dab910a86affd90d888d82f16e9faaf33f5ee61"
+wallet_address_test = "0xeDe4C2b4BdBE580750a99F016b0A1581C3808FA3"
+[chain]
+etherscan_api_key = "2G4MBDIY1RNFHVZS27HR1KKVBHCX8YJ7WE"
+#rpc_url = "https://ethereum-holesky-rpc.publicnode.com"
```
Comment (zk-steve) [inline config.toml]: remove

File: prover-sdk/Cargo.toml [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+sp1-sdk = { git = "https://github.com/succinctlabs/sp1.git", rev = "v1.0.1" }
+aligned-sp1-prover = { workspace = true }
+bincode = { workspace = true }
+hex = { workspace = true }
+aligned-sdk = { workspace = true }
+ethers = { workspace = true }
+anyhow = { workspace = true }
+tokio = { workspace = true }
+ecies = { workspace = true }
+dialoguer = { workspace = true }
```
Comment (zk-steve) [inline prover-sdk/Cargo.toml]: sort

File: sp1-prover/Cargo.toml [hunk 25]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "aligned-sp1-prover"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+sp1-zkvm = { git = "https://github.com/succinctlabs/sp1.git", rev = "v1.0.8-testnet" }
+serde = "1.0.210"
+tiny-keccak = { git = "https://github.com/sp1-patches/tiny-keccak", branch = "patch-v2.0.2", features = ["keccak"] }
+ecies = { version = "0.2.6", default-features = false, features = ["pure", "std"]}
```
Comment (zk-steve) [inline sp1-prover/Cargo.toml]: sort

File: contracts/test/ZkAuction.t.sol
Comment (zk-steve) [inline contracts/test/ZkAuction.t.sol:1]: file name in snake_case

File: contracts/src/mocks/MockToken.sol
Comment (zk-steve) [inline contracts/src/mocks/MockToken.sol:1]: file name in snake_case

File: contracts/src/mocks/MockNFT.sol
Comment (zk-steve) [inline contracts/src/mocks/MockNFT.sol:1]: file name in snake_case

File: contracts/src/ZkAuction.sol
Comment (zk-steve) [inline contracts/src/ZkAuction.sol:1]: file name in snake_case

Comment (zk-steve) [reply contracts/src/ZkAuction.sol:1]: ? Done what

Comment (andrew2003) [reply contracts/src/ZkAuction.sol:1]: Convention correct

File: src/bin/tahken.rs (hunk 16 above)
Comment (andrew2003) [reply src/bin/tahken.rs]: Done

File: src/main.rs (hunk 3 above)
Comment (andrew2003) [reply src/main.rs]: Done

File: Makefile [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+CONFIG_PATH=config.toml
+
+deploy-contract:
+	@chmod +x ./contracts/deploy.sh
+	@./contracts/deploy.sh
+	@cp contracts/out/ZkAuction.sol/ZkAuction.json assets/ZkAuction.json
+
+test-auction:
+	RUST_BACKTRACE=1 cargo test --release --color=always --message-format=json-diagnostic-rendered-ansi --no-run --package zk_auction --lib tests::test_auction_service::test_auction_service --profile dev
```
Comment (zk-steve) [inline Makefile]: Don't copy-paste RustOver command to here 
```suggestion
	cargo test --package zk_auction --lib tests::test_auction_service::test_auction_service --profile dev
```
--no-run ??
Screenshot 2024-10-30 at 11 30 19

File: Makefile [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+	/home/ubuntu/.cargo/bin/cargo test --color=always --message-format=json-diagnostic-rendered-ansi --no-run --package zk_auction --lib tests::test_auction_service::test_auction_service
```
Comment (zk-steve) [inline Makefile]: /home/ubuntu/.cargo/bin/cargo ?

File: Cargo.toml [hunk 28]
```diff
@@ -3,4 +3,49 @@ name = "zk_auction"
...
+curl = "0.4.46"
+dotenv = "0.15.0"
+glob = "0.3.1"
+home = "0.5.9"
+rpassword = "7.3.1"
+serde = { version = "1.0.204", features = ["derive"] }
+serde_derive = "1.0.210"
+serde_json = "1.0.128"
+thiserror = "1.0.63"
+toml = "0.8.19"
```
Comment (zk-steve) [inline Cargo.toml:38]: use workspace for all of these

File: Cargo.toml
Comment (zk-steve) [inline Cargo.toml:1]: no rust code in root folder, see this https://github.com/zk-steve/frog

File: README.md [hunk 29]
```diff
@@ -0,0 +1,163 @@
+# TAHM-KENCH
+
+[![License](https://img.shields.io/github/license/sota-zk-labs/orn)](./LICENSE)
+
+[//]: # ([![Continuous Integration]&#40;https://github.com/sota-zk/orn/actions/workflows/ci.yaml/badge.svg&#41;]&#40;https://github.com/sota-zk/orn/actions/workflows/ci.yaml/badge.svg&#41;)
```
Comment (zk-steve) [inline README.md]: Fix, or remove this

File: README.md [hunk 30]
```diff
@@ -0,0 +1,163 @@
...
+
+## Introduction
+
+**Tahm-Kench** is a [**Sealed-Bid Auction**](https://www.investopedia.com/terms/s/sealed-bid-auction.asp) platform built
+using **SP1 zkVM** and **Aligned**. The project aims to facilitate secure and private auctions by leveraging
```
Comment (zk-steve) [inline README.md]: Links for SP1 and Aligned

File: sp1-prover/src/main.rs (hunk 18 above)
Comment (hduoc2003) [reply sp1-prover/src/main.rs]: the fuction params require reference

File: README.md [hunk 31]
```diff
@@ -0,0 +1,163 @@
...
+### Requirements
+
+Ensure you have the following installed:
+
+1. [Rust](https://www.rust-lang.org/tools/install)
+2. [Foundry](https://getfoundry.sh)
+3. [Aligned CLI](https://docs.alignedlayer.com/introduction/1_try_aligned)
+4. [SP1](https://docs.succinct.xyz/getting-started/install.html)
+
+### Setup
```
Comment (zk-steve) [inline README.md]: Reorganize structure of this file

File: sp1-prover/src/main.rs (hunk 19 above)
Comment (hduoc2003) [reply sp1-prover/src/main.rs]: no, it doesn't

File: README.md (hunk 31 above)
Comment (hduoc2003) [reply README.md]: I wrote this following the required structure of Aligned
image

File: prover-sdk/src/lib.rs [hunk 32]
```diff
@@ -0,0 +1,301 @@
...
+    // let winner_addr = vec![1u8];
+    // let winner_amount = 0; // winner amount
+    // // dbg!(proof.len());
+    // let hash = [202u8, 36, 143, 90, 16, 137, 94, 111, 213, 3, 201, 186, 171, 70, 43, 164, 32, 123, 86, 217, 241, 250, 209, 191, 120, 60, 15, 217, 120, 122, 228, 86];
+    // let addr = [237u8, 228, 194, 180, 189, 190, 88, 7, 80, 169, 159, 1, 107, 10, 21, 129, 195, 128, 143, 163];
+    // let amount = 3u128;
+    // let g = encode(&[Token::FixedBytes(hash.to_vec()), Token::Address(Address::from(addr)), Token::Uint(Uint::from(amount))]);
+    // println!("{}", hex::encode(hash));
+    // println!("{}", hex::encode(addr));
+    // println!("{:?}", g);
```
Comment (zk-steve) [inline prover-sdk/src/lib.rs]: remove

File: prover-sdk/src/lib.rs [hunk 33]
```diff
@@ -0,0 +1,301 @@
...
+
+pub fn flatten(vec: &[[u8; 32]]) -> Vec<u8> {
+    let mut res = vec![];
+    for v in vec.iter() {
+        res.extend_from_slice(v);
+    }
+    res
+}
+
+fn read_file() {}
```
Comment (zk-steve) [inline prover-sdk/src/lib.rs]: ?

File: src/bin/tahken.rs [hunk 34]
```diff
@@ -0,0 +1,242 @@
...
+#[command(about, long_about = None)]
+struct Cli {
+    #[command(subcommand)]
+    command: Option<Commands>,
+    #[clap(short, long)]
+    version: bool,
+    #[clap(short, long, default_value = "config.toml")]
+    config_path: String,
+}
+#[derive(Subcommand, Clone, Debug, PartialEq)]
```
Comment (zk-steve) [inline src/bin/tahken.rs]:
```suggestion

#[derive(Subcommand, Clone, Debug, PartialEq)]
```

Comment (andrew2003) [reply src/bin/tahken.rs]: Complete

File: README.md (hunk 31 above)
Comment (zk-steve) [reply README.md]: Where's "Demo", "Instructions" ???

Comment (zk-steve) [reply README.md]: Roadmap is the last?

File: config.toml (hunk 6 above)
Comment (zk-steve) [reply config.toml:2]: Done what?

File: sp1-prover/taplo/taplo.toml
Comment (zk-steve) [inline sp1-prover/taplo/taplo.toml:1]: multiple taplo files ?

File: prover-sdk/src/lib.rs
Comment (zk-steve) [inline prover-sdk/src/lib.rs:1]: Maybe no need to split this as a separate package

File: Makefile (hunk 27 above)
Comment (andrew2003) [reply Makefile]: It's not necessary. I was removed it.

File: sp1-prover/taplo/taplo.toml
Comment (andrew2003) [reply sp1-prover/taplo/taplo.toml:1]: I was removed it

**Pull request merged**